include/glow/Runtime/HostManager/HostManager.h

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_RUNTIME_HOSTMANAGER_H
+#define GLOW_RUNTIME_HOSTMANAGER_H
+#include "glow/Graph/Graph.h"
+#include "glow/Runtime/Executor.h"
+#include "glow/Runtime/Provisioner.h"
+#include "glow/Runtime/RuntimeTypes.h"
+
+#include <llvm/ADT/StringRef.h>
+
+#include <atomic>
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+namespace glow {
+namespace runtime {
+
+/// Callback signalling the result of running a function.
+using ResultCBTy = std::function<void(
+    runtime::RunIdentifierTy, runtime::ResultCode, std::unique_ptr<Context>)>;
+
+class DeviceManager;
+class Context;
+class DeviceConfig;
+class HostManager final : std::enable_shared_from_this<HostManager> {
+  /// Count of current in-flight networks being run.
+  std::atomic<size_t> activeRequestCount_{0};
+
+  /// Count of total requests, this is used as a run ID.
+  std::atomic<size_t> totalRequestCount_{0};
+
+  /// Limit maximum count of networks run at once. Hardcoded for now this should
+  /// be a configurable value.
+  const unsigned int activeRequestLimit_ = 20;
+  /// A map from a networkID to the DAG that represents the network.
+  std::unordered_map<std::string, DAGNode> networks_;
```

Comment:
Is the NetworkID a string?

Reply:
Why do we need to leak the ONNX-IFI details here? Why do we need network ID?  At this part of the compiler we need to use Module*, Function* and Node*.

Reply:
Agree, I think Module* would be best here. In earlier versions of the HostManager interface, ownership of Module was lost during addNetwork(), this isn't the case any more so a separate ID shouldn't be necessary now.

Reply:
Also, what about concurrency here? I think this needs some locking, otherwise addNetwork, removeNetwork, and runNetwork can't be run concurrently.

Reply:
renaming NetworkId to networkName for clarity, Within a given module there are multiple functions. They each have a unique string name identifying them. Currently the plan is to use that string, unless a different unique identifier is determined by the Caffe2/ONNX-IFI to refer to a network once it's initialized. (This is flexible, we just need  a unique identifier so the HostManager can map that to its DAG which contains the information to execute.) 

I agree on the locking, it needs to be added to allow concurrency.